hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java

```diff
@@ -1182,7 +1183,12 @@ public void sendRpcRequest(final Call call)
       final ResponseBuffer buf = new ResponseBuffer();
       header.writeDelimitedTo(buf);
       RpcWritable.wrap(call.rpcRequest).writeTo(buf);
-      rpcRequestQueue.put(Pair.of(call, buf));
+      synchronized (sendRpcRequestLock) {
```

Comment:
markClosed doesn't use the same lock, so i don't see this stopping a race condition between
this submission and the markClosed set+purge queue;

better: move rpcRequestQueue.put(Pair.of(call, buf)); into a new method, and make synchronized

```java
private sychronized put(pair) {
if (!shouldCloseConnection.get()) {
     rpcRequestQueue.put(pair);
  }
}
```

Reply:
Agree, we should use either of these options to prevent race condition:
1. `sendRpcRequestLock` for both places: here in sendRpcRequest and before markClosed CAS operation
2. Use separate synchronized method for put operation here in sendRpcRequest (being tracked by #5366)

Reply:
Thanks @steveloughran @virajjasani for your review and sorry for my late response.

I considered this trace during coding.  If the trace happened, my modification only remained one call into the `rpcRequestQueue`, it will not caused the sender thread blocked. And the closed connection will be destroyed by GC.

Reply:
```
synchronized (sendRpcRequestLock) {
    if (shouldCloseConnection.get()) {
        return;
     }
      rpcRequestQueue.put(Pair.of(call, buf));
 }
```
About this issue, we only need to ensure that the sender thread will not be blocked by the `rpcRequestQueue`, and we don't care about whether there are residual requests in the  `rpcRequestQueue`. 

This is my coding idea, and looking forward your any good suggestions. @steveloughran @virajjasani